### PR TITLE
Complete the [self class] fold chain: namespace qualifiers/tail const-folds + method-body escaping model (#1096, #1097)

### LINKED ISSUE #1096: `namespace qualifiers` / `namespace tail` have no const_fold (blocks the ticklecharts `${ns}::setdef` chain)
Found during the `[self class]` folding PR (#1095): issue #1080 asserted these "already fold" — verified false. With `[self class]` now folding, the real-corpus chain `set ns [namespace qualifiers [self class]]; ${ns}::setdef …` (ticklecharts; audit idx 44/54's mechanic) still stops after the inner hop because `namespace qualifiers`/`namespace tail` have no fold callback.

The full edge-case behaviour is already oracle-pinned (9.0.4 + 8.6, identical) and captured in #1095's probe transcripts, ready to become the acceptance tests:

| input | `qualifiers` | `tail` |
|---|---|---|
| `:::` | `{}` | `{}` |
| `a:::b` | `a` | `b` |
| `::a::b::` | `::a::b` | `{}` |
| `::x:y` | `{}` | `x:y` |

Fix direction: registry fold callbacks on both subcommands (same `CommandSpec` fold surface `string`/`expr` folds use), string-math only — no namespace existence checks (the commands are pure string operations, per the man page and the transcript). Second blocker for the full chain is tracked separately: variable propagation inside method bodies needs the instance-variable escaping model.

### LINKED ISSUE #1097: Variable propagation in method bodies needs an instance-variable escaping model (propagation.rs vs elimination.rs asymmetry)
Found during the `[self class]` folding PR (#1095). The general SCCP propagation pass deliberately stays OFF for method bodies: an instance variable is object state that outlives the frame, and any `my …` call can rewrite it, so `set ivar 1; my other; puts $ivar` would be miscompiled by the per-function constants lattice. `elimination.rs` already has the escaping model for this; `propagation.rs` does not — #1095 worked around it with a dedicated no-constants-map walk (`run_oo_method_folds`) that folds only `[self class]`.

Consequence: `$var` propagation, `[string …]` folds over method-local variables, and the `${ns}` hop of the ticklecharts chain (`set ns [namespace qualifiers [self class]]; ${ns}::setdef`) don't fold inside methods even when provably safe (a method-local `set` never aliased by instance state).

Fix direction: port/share elimination.rs's escaping classification into the propagation lattice — a variable is propagatable inside a method body iff it's provably method-local (never declared via `variable`, not in the class's variable list, not touched through `my variable`/`upvar`), with `my`/`next`/`[self …]` dispatches acting as barriers only for escaping variables. Abstain-toward-no-fold stays the rule. Blocks the second half of #1096's chain; also unblocks the deliberate narrowing documented in #1095.

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

PR D7 — the two remaining hops of the ticklecharts `${ns}::setdef` chain (#1095's follow-ups), plus an honest finding about what still blocks the navigation half.

- **#1096 — `namespace qualifiers`/`tail` const-folds.** Byte-exact ports of `NamespaceQualifiersCmd`/`NamespaceTailCmd` (tclNamesp.c — backwards colon-run scan, with the leading-`::`-is-not-a-separator subtlety that makes `namespace tail :` answer `:`), wired as registry `const_fold` callbacks on the two `SubCommand`s — the same surface `string`/`dict`/`list` folds use, zero optimiser-side change. A 21-row oracle matrix byte-identical on 9.0.4/8.6.14 (registered version-invariant, and that invariance is itself differential-tested against *both* interpreters), UTF-8-safe by construction, namespace-independence re-confirmed.
- **#1097 — method-body escaping model.** `run_oo_method_folds` now re-runs SCCP with `MethodDef::instance_vars` in the escaping set (re-running, not filtering — a name filter can't poison derived values like `set a $ivar; set b $a`), confined **to the pass**: the first cut widened the shared `FunctionUnit` lattice and the full workspace gate caught it regressing #797's object-collection typing — redesigned so `compilation_unit.rs` is untouched. A whole-module barrier (`has_caller_frame_alias` / `EVALUATES_IN_SHIFTED_FRAME`) bars propagation when any method can reach a caller frame. Plain `$var` O100 propagation now works inside method bodies for the first time.

**Result at the optimiser level:** `set ns [namespace qualifiers [self class]]` folds to the parent namespace in one O129. **The navigation e2e does not close**, and the probe evidence shows why with zero OO involvement: the analyser's constant lattice doesn't fold a constant command-substitution `set` RHS (filed **#1132** — the actual blocker), and `definition()` ignores an already-resolved `indirect` head, jumping to a same-named decoy even for `set ns ::tc; ${ns}::setdef` (filed **#1133** — which partially undercuts audit idx 44's navigation claim; STATUS.md annotated additively, no counts changed). Also filed: **#1134** (SCCP doesn't feed O129 results back into the value lattice for multi-statement chains), **#1135** (two independently-flaky e2e tests making the workspace gate non-deterministic for agents), **#1136** (`make check-rust` broken inside git worktrees via `tcl-spec-studio-wasm` workspace resolution).

## Validation

- [x] `make test-rust` full workspace — every crate green; the one e2e failure across runs is the known settings-timing flake family (each passes in isolation, different test failed on a different run; filed #1135). **The first full run caught the #797 regression** — that's the gate doing its job.
- [x] TP/TN suites: the 21-row registry oracle + UTF-8 + arity-decline TNs; both-dialect differential fold tests; optimiser TP (chain in one O129, method-local propagation) and TN (dynamic args, instance variables via all three declaration forms, the `my bump` miscompile shape, caller-frame barrier).
- [x] clippy `-D warnings` clean, **zero new allows** (the first draft's `isize` casts were rewritten with unsigned arithmetic instead of allowed); fmt clean; xtask-check all OK.

## Compiler/KCS checklist (when touching compiler behaviour, diagnostics, or analysis contracts)

- [x] Did this change alter a compiler fact contract? Yes — two new registry const-folds; method bodies now propagate provably-local constants under the documented escaping barrier.
- [x] KCS notes updated: O129 folding page (namespace string-ops section), `sccp-core-analyses.md` (the method-body escaping model + barrier).
- [ ] New compiler KCS note linked. (n/a — existing notes updated)

Closes #1096. Closes #1097.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Mf12569PiJ3F78PAp2iVX7

### CODE REVIEW

File: rust/tcl-compiler/src/optimiser/propagation.rs [hunk 1]
```diff
@@ -909,22 +909,126 @@ fn oo_frame_for(
...
+    };
+    if fu.complexity_guarded {
+        return std::collections::HashMap::new();
+    }
+    let sccp = crate::sccp::sccp_with_extra_escaping(
+        &fu.cfg,
+        &fu.ssa,
+        None,
+        FoldPolicy::from_registry(registry),
+        &method.instance_vars,
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/optimiser/propagation.rs:1019]: Confirmed and fixed in 4ad74f729 by the merge, as suggested. Reproduced exactly this shape: pre-fix the pass emits an `O100` replacing `$x` with `1`, where tclsh 9.0.4 and 8.6.14 both print `2`.

Root cause is as described — `extract_oo_methods` builds a fresh `class_ivars` set per definition block (lowering/mod.rs), and the `contains_key` → `continue` guard means a method extracted from an earlier block is never revisited. A block that declares *only* `variable x` and no methods contributed nothing at all.

Fixed at lowering time with a per-class union: `Lowerer::class_instance_vars` accumulates `class qualified name -> declarations` across every definition block, and `extract_oo_methods_pass` merges it into every method of that class once all blocks are walked. Order-free by construction — declaring `variable x` anywhere makes it instance state for every method of the class — and pinned by a test that runs the same program with the `oo::define`-first ordering.

Worth noting the blast radius is wider than the optimiser: `elimination.rs` feeds the same `MethodDef::instance_vars` through its cross-event channel to stop an instance-state write being deleted as a dead store, so it had the identical hole and is fixed by the same change.

Test: `instance_vars_declared_in_a_later_definition_block_bar_propagation`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rust/tcl-compiler/src/optimiser/propagation.rs [hunk 2]
```diff
@@ -909,22 +909,126 @@ fn oo_frame_for(
...
+/// `method b {} {upvar 1 ns n; set n Y}` would silently mis-propagate.
+/// Abstain-toward-no-fold: when any method body could do it, no method body's
+/// locals are propagated.
+///
+/// Flow-insensitive and whole-module for the same reason
+/// [`crate::command_binding::ModuleCommandMutations::trusts`] is: the dispatch
+/// order between methods is not statically known.
+fn methods_reach_caller_frames(cu: &CompilationUnit, registry: &CommandRegistry) -> bool {
+    use tcl_registry::Traits;
+    cu.methods.values().any(|fu| {
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/optimiser/propagation.rs]: Confirmed and fixed in 4ad74f729. Reproduced: an initially empty `helper`, later redefined as `{upvar 1 x y; set y 2}`, pre-fix lets `$x` fold to `1`; tclsh prints `2` on both 9.0.4 and 8.6.14.

I checked what `redefined_methods` actually stores before scoping the switch: it is a `HashSet<String>` of method *qualified names*, and the replacement **body is discarded** (the `continue` fires before the insert), so there is nothing extra to scan — abstention is the only option.

On scoping: I went with the whole-module kill switch rather than per-class, and I think per-class would be unsound rather than merely imprecise. `my` dispatches along the MRO, so a method replaced in a *superclass* is reachable from a subclass's body:

```tcl
oo::class create B { method helper {} {} }
oo::class create C { superclass B; method m {} {set x 1; my helper; puts $x} }
oo::define B { method helper {} { upvar 1 x y; set y 2 } }
```

Scoping to `C` misses this. Getting it right per-class would need the MRO, which this pass does not compute. Redefinition is rare, so the precision cost is small and the soundness argument does not depend on machinery that is not here. Both shapes are tested.

Test: `a_redefined_method_bars_propagation`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rust/tcl-compiler/src/optimiser/propagation.rs [hunk 3]
```diff
@@ -909,22 +909,126 @@ fn oo_frame_for(
...
+        if crate::var_observability::analyse_var_observability(&fu.cfg, registry)
+            .has_caller_frame_alias()
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/optimiser/propagation.rs]: Confirmed and fixed in 4ad74f729, and you were right that the fix belongs at the evidence source rather than in the skip list. Reproduced this shape verbatim; pre-fix `$x` folds to `1`, tclsh prints `2`.

I checked whether `VarObservability`'s own alias indices already covered it: they do not — `stmt_gen`'s `EscapeFlag::UPVAR` mark goes through the *same* `upvar_local_declaration_indices`, so the gap was in the shared helper, not just the `stmt_gen` route. Patching that skip list locally would have been wrong anyway: other consumers rely on it to *name* which local is aliased, and a dynamic pair genuinely has no name to give them.

So I switched the gate to a complete source instead. `cfg_builder::upvar_info::collect_upvar_targets` — the same summary `detect_upvar_procs` builds for procs — already classifies this case correctly (`$src` is a parameter, so it lands in `param_targets`; a non-parameter `$n` sets `has_unresolvable_caller_target`). The new `reaches_caller_frame` wraps it as the structural question "can this body reach its caller's frame at all?", and the gate calls that. Methods and procs now share one evidence source.

That review turned up one adjacent hole worth flagging: `record_upvar_call` also drops a pair whose **local** side is dynamic (`upvar 1 x $dst`) — it `continue`s because there is no key to file the alias under — so that shape was invisible too. Added `UpvarInfo::unnameable_local_aliases` recording the caller-side word, deliberately outside `is_empty()`'s summary contract and read only by `reaches_caller_frame`, so no caller-side resolver changes behaviour. All three shapes are tested.

`has_caller_frame_alias` is deleted from `var_observability` (it was mine, added in this PR, and is fully superseded). The paired `EVALUATES_IN_SHIFTED_FRAME` scan goes with it — `UpvarInfo` models `uplevel` levels more precisely, so `uplevel #0`, which cannot touch the caller, no longer over-fires.

Tests: `a_substituted_upvar_source_counts_as_a_caller_frame_alias`, plus `a_namespace_alias_is_not_a_caller_frame_alias` guarding the other direction — `global`/`variable`/`namespace upvar` reach a namespace, not a caller frame, and must not trip the barrier or every ordinary class body stops propagating.

---
_Generated by [Claude Code](https://claude.ai/code)_
